crates/storage/provider/src/post_state.rs

```diff
@@ -278,52 +177,87 @@ impl PostState {
         calculate_receipt_root_ref(self.receipts().iter().map(Into::into))
     }
 
-    /// Get the number of transitions causing this [PostState]
-    pub fn transitions_count(&self) -> TransitionId {
-        self.current_transition_id
-    }
-
+    // todo: note overwrite behavior, i.e. changes in `other` take precedent
     /// Extend this [PostState] with the changes in another [PostState].
-    pub fn extend(&mut self, other: PostState) {
-        if other.changes.is_empty() {
-            return
+    pub fn extend(&mut self, mut other: PostState) {
+        // Update plain state
+        self.accounts.extend(other.accounts);
+        for (address, their_storage) in other.storage {
+            let our_storage = self.storage.entry(address).or_default();
+            if their_storage.wiped {
+                our_storage.wiped = true;
+                our_storage.storage.clear();
+            }
+            our_storage.storage.extend(their_storage.storage);
         }
 
-        self.changes.reserve(other.changes.len());
+        // Insert account change sets
+        for (block_number, account_changes) in std::mem::take(&mut other.account_changes) {
+            let block = self.account_changes.entry(block_number).or_default();
+            for (address, account) in account_changes {
+                // TODO: Note about why we skip accounts we already have
+                if block.contains_key(&address) {
+                    continue
+                }
```

Comment:
In case you have 2 txs w/ PostStates you always want to keep the earliest value for the changeset, so if you have tx0 and tx1 and they both changed the same account, the view from tx1 would be different from tx0, and we want to keep the changeset from tx0 instead.

Reply:
Yeah, we want the oldest values that the block has seen.